File: gray-code.cc
#include <vector>
using namespace std;

class Solution {
public:
  vector<int> grayCode(int n) {
    vector<int> ans;
    ans.push_back(0);
    int total = 1;
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < total; j++) {
        ans.push_back(ans[j] | (1 << i));
      }
      reverse(ans.begin() + total, ans.end());
      total *= 2;
    }
    return ans;
  }
};

File: gray-code.java
public class Solution {
    public ArrayList<Integer> grayCode(int n) {
        ArrayList<Integer> codes = new ArrayList<Integer>();
        codes.add(0);
        for (int i = 0; i < n; i++) {
            int size = codes.size();
            for (int j = size - 1; j >= 0; j--)
                codes.add(codes.get(j) | (1 << i));
        }
        return codes;
    }
}
